### PR TITLE
Set variables for CakePHP and Inertia components

### PR DESCRIPTION
**Pull request by toggenation**
Hi Ishan,

I barely know how to submit a PR so I hope I have done this right. Please let me know how I can do it better if this is not right.

I have added the ability to specify viewVars that should not be included in the inertia div and instead be available to whichever layout/template is being served by InertiaWebView.

This allows integrating the inertia part of a page with a legacy navbar / menu.

### CODE REVIEW

**Comment by LordSimal** [comment]
To make Github Actions happy you should stick with the configured coding style 😉 
This can be automatically fixed by executing `composer cs-fix`.
But to be 100% sure you should then also execute `composer cs-check` and see if something needs to be manually fixed.

For the PHPStan problems you can either look at github actions here https://github.com/ishanvyas22/cakephp-inertiajs/actions/runs/4838281986/jobs/8622646318?pr=20#step:9:52 or execute `composer analyse`
These issues can't be fixed automatically so you have to follow the problems description 😁

**Comment by ishanvyas22** [comment]
Hey @toggenation, thanks for the PR.

I looked through the code and the use case you suggested. I'm not sure we will need changes in the code to make this "non-inertia" thing work. What if you customize the `beforeRender` method in the controller, in that you first call `beforeRender` of `InertiaResponseTrait`([more on customizing `beforeRender` hook](https://github.com/ishanvyas22/cakephp-inertiajs/blob/master/docs/ServerSideSetup.md#customize-beforerender-hook)) and then set view vars that you want as "non-inertia".

I haven't tested this approach but if this works we might be able to document it so users know.

**Comment by toggenation** [comment]
> 

Hi Ishan,
I tried the following `beforeRender` code in my Controller but because `InertiaWebView::render` is called after this customized `beforeRender` code it still grabs all the "non-Inertia" vars and `sets` them into `page`. 

Am I missing something?

```php
class DespatchController extends AppController
{
    use ErrorFormatterTrait;

    // Alias the trait's `beforeRender` method
    use InertiaResponseTrait {
        beforeRender as protected inertiaBeforeRender;
    }

    public function beforeRender(EventInterface $event)
    {
        $this->inertiaBeforeRender($event);

        // Do your customization here.

        $this->viewBuilder()
            ->setLayout('Inertia/default');

        $currentVars = $this->viewBuilder()->getVars();

        $nonInertiaVars = [
            'companyName',
            'showAsn',
            'showLocation',
            'menuTree',
            'isAdmin',
            'user',
        ];

        $currentVars = array_intersect_key($currentVars, array_flip($nonInertiaVars));

        foreach ($currentVars as $var => $value) {
            $this->set($var, $value);
        }

        // this doesn't work either
        // $this->viewBuilder()->setVars($currentVars);

    }
```

**Comment by ishanvyas22** [comment]
Alright. Then I don't prefer settings "non-inertia" vars like this, because setting the vars twice(initially as regular and again specifying as `nonInertiaVars`) seems a bit redundant to me.

One approach I can think of now is setting a `$_nonInertiaProps` (feel free to suggest better name) array inside the controller:

```php
class ExampleController extends AppController
{
    use InertiaResponseTrait;

    protected $_nonInertiaProps = ['menuTree', 'user', ...];

    ...
}
``` 

Then this array can be used to filter out view vars from inertia props array.

@toggenation What do you think about this approach?

**Comment by toggenation** [comment]
> @toggenation What do you think about this approach?

Hi @ishanvyas22, I think it's a better approach than anything I've cobbled together.

I have made a change to the code. After setting `$_nonInertiaProps` in the controller. 

Modify the `InertiaResponseTrait` to check for `$_nonInertiaProps` and set it as a view config option. 

Then in `BaseViewTrait::getProps` filter out the props specified in `$_nonInertiaProps` 

The specified vars are passed through to CakePHP and not wrapped by `$page = { props => $this->getProps() }`

I have murdered my `git` history by ignorance so I apologize if it makes it difficult for you. Let me know how I could do better.

File: src/View/BaseViewTrait.php [hunk 1]
```diff
@@ -75,14 +82,45 @@ private function getProps(): array
...
 
+    /**
+     * Any viewVars that should not be included in Inertia `page`
+     * are set here
+     *
+     * @param mixed $props The view vars
+     * @param mixed $nonInertiaVars Array of non Inertia Vars
+     * @return array
+     */
+    private function setNonInertiaVars($props, $nonInertiaVars): array
```
Comment (ishanvyas22) [inline src/View/BaseViewTrait.php]: Is it used? I don't see it used anywhere so you can safely remove it.

File: src/View/BaseViewTrait.php [hunk 2]
```diff
@@ -53,13 +53,20 @@ private function getProps(): array
...
 
-        $this->viewVars = [];
+        $nonInertiaProps = $this->getConfig('_nonInertiaProps') ?? [];
+
+        $onlyViewVars = array_diff($onlyViewVars, $nonInertiaProps);
+
+        $this->viewVars = array_intersect_key(
+            $passedViewVars,
+            array_flip($nonInertiaProps)
+        );
```
Comment (ishanvyas22) [inline src/View/BaseViewTrait.php:67]: Can you describe what is this piece of code doing? I.e. add comment why it's there and brief meaning of what code is doing.

File: src/View/BaseViewTrait.php [hunk 3]
```diff
@@ -53,13 +53,20 @@ private function getProps(): array
...
         $props = [];
         $only = $this->getPartialData();
-        $onlyViewVars = ! empty($only) ? $only : array_keys($this->viewVars);
+        $onlyViewVars = !empty($only) ? $only : array_keys($this->viewVars);
         $passedViewVars = $this->viewVars;
 
-        $this->viewVars = [];
+        $nonInertiaProps = $this->getConfig('_nonInertiaProps') ?? [];
+
+        $onlyViewVars = array_diff($onlyViewVars, $nonInertiaProps);
```
Comment (ishanvyas22) [inline src/View/BaseViewTrait.php]: I would suggest to move these lines of code into a separate (private) method.

For example:
```php
        $onlyViewVars = $this->filterViewVars();
```

File: tests/TestCase/View/InertiaWebViewTest.php [hunk 4]
```diff
@@ -38,4 +41,51 @@ public function testRendersComponentName()
...
+    public function createViewBuilderForPassThru()
+    {
+        $request = new ServerRequest();
+        $response = new Response();
+
+        $this->View = (new ViewBuilder())
+            ->setClassName('Inertia\View\InertiaWebView')
+            ->setOption('_nonInertiaProps', ['one', 'two', 'three'])
+            ->build([], $request, $response);
+    }
```
Comment (ishanvyas22) [inline tests/TestCase/View/InertiaWebViewTest.php]: `setUp` method is already creating the view builder class, why you need to do it again?

**Review by ishanvyas22 — CHANGES_REQUESTED** [review]
Looking in a bit better shape now. Dropping few improvement points.

In addition to those comments, we still have to update the docs. So if you can write a bit about it as earlier you did that would be great :)

File: src/View/BaseViewTrait.php (hunk 1 above)
Comment (toggenation) [reply src/View/BaseViewTrait.php]: Will get rid of that

File: tests/TestCase/View/InertiaWebViewTest.php (hunk 4 above)
Comment (toggenation) [reply tests/TestCase/View/InertiaWebViewTest.php]: Hi, I needed the ViewBuilder to set the `_nonInertiaProps` option as it doesn't effect the other tests I've moved it to `setUp`

**Review by ishanvyas22 — APPROVED** [review]
Looks good

**Comment by ishanvyas22** [comment]
Released with [2.3.0](https://github.com/ishanvyas22/cakephp-inertiajs/releases/tag/2.3.0). Apologies for the delay.

**Pull request merged**